### PR TITLE
Added forms documentation

### PR DESCRIPTION
**Pull request by stacymcauliffe**
Added forms documentation per issue [#389](https://github.com/patternfly/patternfly-next/issues/389) in the pf-next repo

Closes https://github.com/patternfly/patternfly-next/issues/389

### CODE REVIEW

**Comment by jgiardino** [comment]
@seanforyou23 and I have been reviewing the Deque University course materials for forms. I'm still working my way through this document and their course materials, which is taking some time, but wanted to go ahead and provide feedback about the "**Provide the right kind of help**" section.

It's pretty amazing how much this document lines up with what Deque recommends! You nailed it with this statement about Field Level Help: "Never hide critical information in a popover."

Here are some recommendations on this section:
- Deque advises that any information that's critical to completing an input field should be visible. Just like you have a DON'T statement about not putting critical info in a popover, we should also add a DO statement about putting critical information in the Helper Text (i.e. visible text below the field).
- There's a major difference with the placement of our Field Level Help button and what they recommend. Their advice is that this button should precede the field input.
  - I'm copying @mcarrano and @mceledonia on this one, since I don't think this is an issue with the documentation in this PR, but more with the design itself. I'm happy to share more details about this, but their point is that having the help text button come after the input field is a major usability issue for screen reader users, since they won't realize it's available until after they've tried to complete the field.
- I feel like the Deque course materials were lacking in information about placeholder text. They basically just summed it up as, don't use placeholder text as labels.  
  - @seanforyou23  talked about what they did have, and these are additional statements that we thought would be good to include for the Syntax Hints:
    - Given the accessibility issues associated with placeholder text, a form should still be usable even if placeholder text is removed.
    - Use “e.g.” in front of placeholder text, since it’s not obvious in some screen readers that placeholder text isn’t the inputted value. 
  - And for the design, it would be nice to consider increasing the contrast of this text, which shouldn't be an issue if we prepend all syntax hints with "e.g. ". Although this isn’t critical either assuming we really emphasize the guideline that placeholder text shouldn’t be critical.

**Comment by stacymcauliffe** [comment]
@jgiardino - I updated the docs to include the example you mentioned. For the placeholder text, the only type of placeholder text we're recommending is syntax hints and avoiding examples with info like _e.g. My Project Name_. I'm not sure and eg. prefix would make sense for a syntax hint, but maybe? If we want to change this, I can update the image. 

@mcarrano let me know if you want to see any additional changes to this PR. Thanks!

**Comment by jgiardino** [comment]
> For the placeholder text, the only type of placeholder text we're recommending is syntax hints and avoiding examples with info like e.g. My Project Name.

Oh yeah, I definitely see that now that I look back at the examples that are included. It might be helpful to be more explicit about what we mean by "syntax hint" and/or to include some common examples in this section to better clarify what we're suggesting. Like a syntax hint for date might be MM/YYYY, and for phone number might be XXX-XXX-XXXX. If this is what we're recommending, then I think it's pretty obvious these are placeholders. I agree the "e.g." might be weird and feels more like a bandaid for a wonky web feature.

File: patternfly-4/components/forms/design.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
+# Forms
+## About
+A form is a group of elements used to collect information from a user. Forms can be presented to the user in a variety of contexts depending on the use case and scenario.
+
+## Elements of a form
+While every form will require a different combination of components depending on the use-case, most forms will contain the following basic elements:
+![Elements of a form](img/form-elements.png)
+1. [Labels](../data_input) - Field labels can be aligned to the left or top of the field depending on the layout of your page and the amount of space you have to work with. Indicate all required text and input fields with a red asterisk. Learn more about field labels in the [text and data input design guidelines](../data_input).
+
+    *Required field*
```
[the remark this answers is not in the document]
Comment (stacymcauliffe) [reply patternfly-4/components/forms/design.md]: I updated the vertical to top. I'm honestly not sure why the images are appearing above the labels in the preview. They're displaying normally in my markdown preview.

File: patternfly-4/components/forms/design.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    <img src= "img/required-field.png" width="300" height="auto" title="Required field">
+
+  *Vertical alignment*
+  <img src= "img/vertical-alignment.png" width="300" height="auto" title="Vertical alignment">
+
+  *Left alignment*
+  <img src= "img/left-alignment.png" width="300" height="auto" title="Left alignment">
+
+2. [Text field](../data_input) - provides an area for users to input free-form text.
+3. [Data input](../data_input) - provides a way for users to submit any information that is not free-form text. Input components are designed for specific use cases.
```
[the remark this answers is not in the document]
Comment (stacymcauliffe) [reply patternfly-4/components/forms/design.md]: Done

File: patternfly-4/components/forms/design.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+* Data input
+
+**Related components**
+* Modal
+* Wizard
+
+## Usage
+### Use progressive disclosure
+
+<img src= "img/progressive-disclosure.png" width="700" height="auto" title="Progressive disclosure">
```
[the remark this answers is not in the document]
Comment (stacymcauliffe) [reply patternfly-4/components/forms/design.md]: Yep, great catch, thanks! Updated the image

File: patternfly-4/components/forms/design.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+* A user leaving a required field incomplete
+
+> Note: Inline error messages will replace helper text.
+
+**Server-side validation**
+Server-side validation is performed when the user attempts to submit a form. In this scenario, the user completes the form in its entirety and the page is reloaded with any errors detected.
+
+For server-side validation, use an inline notification along with inline errors wherever possible to make it simple for the user to quickly identify areas that need to be fixed.
+
+Notification messages should describe the error and help users solve the problem. Once form errors are resolved and submitted, the notification should be replaced with a confirmation message.
```
[the remark this answers is not in the document]
Comment (stacymcauliffe) [reply patternfly-4/components/forms/design.md]: Done

File: patternfly-4/components/forms/design.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+<img src= "img/validation.png" width="300" height="auto" title="validation">
+
+
+### Using forms in your designs
+**Full page**
+![Full page form](img/full-page-example.png)
```
[the remark this answers is not in the document]
Comment (stacymcauliffe) [reply patternfly-4/components/forms/design.md]: I updated the text on this image. I think it helps a bit, but to your point, I'm sure we could some up with a better example

**Comment by stacymcauliffe** [comment]
Made all recommended updates and ready for another review.

File: patternfly-4/components/forms/design.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcarrano) [reply patternfly-4/components/forms/design.md]: That seems to have fixed the problem.

**Comment by mcarrano** [comment]
We have two +1's so I am going to proceed with a merge. Thanks again for your effort @stacymcauliffe

**Pull request merged**